### ISSUE TITLE
Splinter Cell: Double Agent resolution scaling not working properly in single player

### ISSUE THREAD

**Issue by ghost** [xemu-project/xemu#949]
### Title

https://xemu.app/titles/5553005e/#Tom-Clancy-s-Splinter-Cell-Double-Agent

### Bug Description

https://user-images.githubusercontent.com/103341098/169669277-70ecf17f-5d8c-4a28-85a3-70b1d45966db.mp4

Setting the resolution scaling to 2x and onward results in a very aliased image. In this video specifically the resolution scaling was set to 4x. 

### Expected Behavior

https://user-images.githubusercontent.com/103341098/169669515-64271149-64f3-4a59-a042-85039ca10dd3.mp4

This is how that same area looks on Xbox One which has a higher res than OG Xbox.

Strangely, this issue is not present in the coop mode as can be seen in these screenshots, where again the game ran at 4x resolution scaling. Nor is it an issue in Chaos Theory despite both games running on the same engine. 
![xemu exe_Screenshot_2022 01 02_-_23 57 22 71](https://user-images.githubusercontent.com/103341098/169669567-74c5b3c6-343e-4b7f-a31b-f9e4c61db294.png)
![xemu exe_Screenshot_2022 01 03_-_00 08 13 01](https://user-images.githubusercontent.com/103341098/169669569-3e9cec57-e824-4814-8529-ceacf2d80122.png)
![xemu exe_Screenshot_2022 01 03_-_00 14 08 25](https://user-images.githubusercontent.com/103341098/169669570-dd2d0aee-175e-4e0e-b6dc-f77be200c659.png)

Edit: This issue corrects itself whilst using the night vision mode 

No night vision: Image appears aliased 
![xemu exe Screenshot 2022 05 18 - 01 19 46 79](https://user-images.githubusercontent.com/103341098/170154426-1b6fdfd5-42cc-4fca-8dd6-fd202f69c98c.png)

Night vision on: The image appears sharp as can be seen by the wall on the right side and by the metal grate at the end of the hall
![xemu exe Screenshot 2022 05 18 - 01 19 52 44](https://user-images.githubusercontent.com/103341098/170154471-fbcc96ca-2383-4b2f-85b2-97a24ec400c8.png)

### xemu Version

v0.7.17

### System Information

Windows 10
Ryzen 5 3600x
RTX 2060
NVIDIA Driver: 30.0.15.1215

**Comment by Joris-truly** [comment]
Tested it. As of today, the res not scaling is still an issue.

**Comment by JordanLeich** [comment]
> Tested it. As of today, the res not scaling is still an issue.

I tested as well and same result, resolution scale doesnt really work (makes the game look worse + lots of fps loss) and the game still has the border issue as seen in the screenshot.

![image](https://user-images.githubusercontent.com/57107961/218273242-0cb114e3-0e30-4631-ace6-678c98cd865e.png)

**Comment by jcolell** [comment]
It seems to be the same MLAA issue some games have with the RPCS3 Emulator:

https://illusion0001.com/patches/2021/06/25/t1-ps3-mlaa-fix/
https://github.com/RPCS3/rpcs3/issues/10316
https://www.neogaf.com/threads/rpcs3-mlaa-disabled-4k-exclusives.1568413/

The game had an MLAA for improving image quality in single-player mode. Absent in multiplayer (or in night vision mode) for performance reasons. 

So the problem here is the AA and Bloom effect \ shader used in the game.

**Comment by xthedom** [comment]
why has this never been fixed lmao this emu has slow progression

**Comment by BiatuAutMiahn** [comment]
> why has this never been fixed lmao this emu has slow progression

Lol, you must have never known Cxbx. `:P` once mborg and team started working on this, OG Xbox was no longer a console that was considered to be impossible to emulate. Like we had working 360 emulation with Xeina, rpcs3 for PS3, cemu etc. for modern consoles. 

Whoever thought to use Qemu to literally virtualize the Xbox at the hardware level really hit the nail on the head.

**Comment by AbbieDoobie** [comment]
Just curious; could the MLAA potentially be disabled with an AR code (or something like that), as a workaround? Or straight up modifying the game itself?

**Comment by OneShot95** [comment]
> Just curious; could the MLAA potentially be disabled with an AR code (or something like that), as a workaround? Or straight up modifying the game itself?

This is something I'm going to look into, there must be a way to do this.

**Comment by AbbieDoobie** [comment]
There is a workaround using Reshade shader toggles, posted in the Xemu Discord. I am not sure if links are ok, but if so, [here](https://discord.com/channels/680221390359887933/680221390359888154/1217172973975376094) is a direct message link.

**Comment by QmwJlHuSg9pa** [comment]
> There is a workaround using Reshade shader toggles, posted in the Xemu Discord. I am not sure if links are ok, but if so, [here](https://discord.com/channels/680221390359887933/680221390359888154/1217172973975376094) is a direct message link.

Do you mind attaching a screenshot? I can't access Discord.

**Comment by AbbieDoobie** [comment]
I pulled the text directly. **All credit to 'fellow_demo' on the Xemu discord.**

"I injected Reshade with addon support, got Shader Toggler addon and went thru game's shaders and got some results:

1. I found shaders responsible for AA that makes resolution scale look pixelated and turned it off with some good results(there still some areas where it breaks, tho it breaks even for night vision which is usually unaffected by this issue during scaling)
2. found bloom shader that causes green glowing edges and toggle it off (at expanse of bloom obviously, but since reshade already injected, you can add own bloom). one toggle for this shader makes thermal vision partially broken, other toggle breaks night vision, makes shadows pixelated 
3. also made some toggles for shader that makes shadows levels brighter, noise for night vision and HUD toggle

Using this can make some loading screens corrupted, but game loads fine.
P.S. Also for this to work between sessions correctly gotta wipe Shaders in AppData/Xemu/xemu/Shaders or/and turn caching off"

Here is the reshade ini they included: [ShaderToggler.zip](https://github.com/xemu-project/xemu/files/15300790/ShaderToggler.zip)

You will need to use the full install of Reshade, not the standard. You'll know it's the correct version if it has a warning about anticheat. The full version has shadertoggle included.

In my experience this does work, though it is imperfect.

**Comment by VinceIP** [comment]
Just to add to the above comment - this method works great. Just make sure when you install Reshade, you select the Xemu executable, select the OpenGL API, and select the "Shader Toggle" addon in the extras tab within the installer. Put the ShaderToggler ini file in your Xemu directory. Launch Xemu, press Home to open the Reshade menu, select the addons tab, and explore the Shader Toggle addon. Read the instructions there carefully. Specifically, make sure to toggle the AA and Bloom shaders to be applied on startup and press S to enable them immediately. You'll know it's working when your rendering resolution increases and the bloom border disappears. You'll need to use your numpad controls to cycle through some pixel or vertex shaders. You can easily add Reshade bloom to compensate for the lack of in-game bloom, and in fact, I think it looks better than what is normally there. Same for anti-aliasing.

**Comment by jcolell** [comment]
I can confirm that the AA and Bloom bug is gone with the Shader Toggler workaround. However, night vision still has some issues (effect missing). I hope the Xemu team injects this workaround in some way into the emulator
![Screenshot 2024-06-16 223119](https://github.com/xemu-project/xemu/assets/80860708/41d1b607-af3d-49c1-99be-e7bb15e97930)

**Comment by kgalkin92** [comment]
@AbbieDoobie I applied what you mentioned here, but I faced this issue with lighting 
![image](https://github.com/user-attachments/assets/d790b9a3-01e3-43d5-98a4-8acda3c9a19c)

**Comment by Davometric** [comment]
The above shader toggler ini did absolutely nothing for me. Had to make my own. For anyone that isn't having luck with the above ini file, try my preset. 

'Capslock' to remove the border.
'A' to increase resolution scaling (if you're going higher than native, this will fix the pixelation but cutscenes will be messed up so disable it when cutscenes occur).

When installing reshade from their website, you MUST download reshade will FULL ADDON support. Also make sure when you install it, choose xemu.exe and openGL as well as my preset ini. 

When it asks you for addons, select shadertoggler by otis.
Put shadertoggler.ini in the same folder as xemu.

[ShaderToggler.zip](https://github.com/user-attachments/files/18548071/ShaderToggler.zip)

**Comment by mborgerson** [comment]
Please retest with latest

**Comment by JordanLeich** [comment]
> Please retest with latest

Issue still exists on latest Xemu version - 0.8.39

![Image](https://github.com/user-attachments/assets/21e23c15-5732-46a3-9b70-517b05970af0)

**Comment by r2bir2** [comment]
Check my Tutorial: https://www.reddit.com/r/Splintercell/comments/1jsw3bc/updated_splinter_cell_double_agent_v2_xbox_on_pc/- Inspired from this post.
I fell that every few releases shaders will be changed. But you can setup shadertoggler by yourself :) Dont forget to disable caching shaders

**Comment by jcolell** [comment]
> Check my Tutorial: https://www.reddit.com/r/Splintercell/comments/1jsw3bc/updated_splinter_cell_double_agent_v2_xbox_on_pc/- Inspired from this post. I fell that every few releases shaders will be changed. But you can setup shadertoggler by yourself :) Dont forget to disable caching shaders

Helpful guide as a workaround for now, thanks. Now it's not a perfect experience, but playable. Hopefully, the emulator will soon have these fixes natively.

**Comment by Finoderi** [comment]
Resolution scaling isn't working still and the 'Pixelperfect' shader has been broken since 0.8.58.0 version. On 1x resolution the 'green border' doesn't appear anymore.

**Comment by Finoderi** [comment]
I've modified ini-file for ShaderToggler, so disabling those two 'broken' shaders is working again.
I made two versions - one for disabling both shaders at the same time, and another one with two toggles. 'Pixelation' toggle is active at startup in both configs.

[ShaderToggler.zip](https://github.com/user-attachments/files/20851651/ShaderToggler.zip)

**Comment by jcolell** [comment]
I finished the game with the Shader Toggler fix. Overall, for 80% of the game, it works well as a workaround. Anyway, in scenes with excessive lighting effects (like during fire scenes), the fix does not work, probably because the game uses too much MLAA or the resolution drops significantly, too much even for the "sensitivity" of Shader Toggler. I hope the emulator can permanently resolve this MLAA/Resolution and Bloom issue in the future.

**Comment by Finoderi** [comment]
> In scenes with excessive lighting effects (like during fire scenes), the effect does not work...

May be for particle effects AA, or whatever it is, is hardcoded. I tried to find a vertex or pixel shader which 'pixelates' steam but to no avail.
More specifically, it's not the steam or fire itself that causes this issue, but the 'heat haze' effect around it.

**Comment by TrayZorXx67** [comment]
> J’ai modifié le fichier ini pour ShaderToggler, donc la désactivation de ces deux shaders « cassés » fonctionne à nouveau. J’ai fait deux versions - une pour désactiver les deux shaders en même temps, et une autre avec deux bascules. La bascule 'Pixelisation' est active au démarrage dans les deux configurations.
> 
> [ShaderToggler.zip](https://github.com/user-attachments/files/20851651/ShaderToggler.zip)

it's always with reshade in opengl or vulkan

**Comment by Finoderi** [comment]
Here is modified ShaderToggler.ini that works with 0.8.84 version of xemu...

[ShaderToggler.zip](https://github.com/user-attachments/files/20943858/ShaderToggler.zip)

**Comment by Finoderi** [comment]
Same for 0.8.85.

[ShaderToggler.zip](https://github.com/user-attachments/files/20963785/ShaderToggler.zip)

**Comment by stoggs01** [comment]
I can't get the shadertoggler to fix the pixalation issue, can someone assist me?  I followed the instructions to a T.

**Comment by Finoderi** [comment]
What do you mean? ShaderToggler.ini changes nothing in the latest version of xemu or you tried to find shaders to toggle off yourself and couldn't?
If it's the former, shader hashes usually change every time new build is released and ShaderToggler uses hashes to identify shaders. 

Making a new version of ShaderToggler.ini every time a new build is published looks like a rather pointless endeavour ) 

You can try to use ShaderToggler.ini made for the specific version of xemu or find shaders to switch off by yourself following instructions on ShaderToggler github page.

**Comment by stoggs01** [comment]
Do you have a link that page so I can give that a try?

**Comment by Finoderi** [comment]
https://github.com/FransBouma/ShaderToggler

There are two shaders in SCDA related to 'pixelation effect' - one is pixel, the other is vertex. You don't want to toggle off the pixel shader, because it'll mess up cutscens completely.

**Comment by r3538987** [comment]
Wanted to play this one, and tried to fix mentioned issues with LLM (Codex).
User on Discord gave a hint:
`MRArthur: Pretty easy it known issue if you manually change pixel offset you can have correct scaling at whatever res.`
And after some trial and error, something decent came out.
---> **Vulkan only.** 
Because custom binaries are not allowed, we generated patch file which might help you build your own, or maybe push similar changes to upstream. 

///

<img width="2560" height="1410" alt="Image" src="https://github.com/user-attachments/assets/5391218c-89fd-404d-b477-27686ffad9a8" />

Patch update 2. 

```diff --git a/hw/xbox/nv2a/pgraph/vk/blit.c b/hw/xbox/nv2a/pgraph/vk/blit.c
--- a/hw/xbox/nv2a/pgraph/vk/blit.c
+++ b/hw/xbox/nv2a/pgraph/vk/blit.c
@@ -129,6 +129,14 @@
     hwaddr dest_offset = image_blit->out_y * context_surfaces->dest_pitch +
                          image_blit->out_x * bytes_per_pixel;
 
+    if (image_blit->operation == NV09F_SET_OPERATION_SRCCOPY &&
+        context_surfaces->color_format == NV062_SET_COLOR_FORMAT_LE_A8R8G8B8 &&
+        source + source_offset == dest + dest_offset &&
+        context_surfaces->source_pitch == context_surfaces->dest_pitch) {
+        // Keep the scaled surface; the guest RAM copy was synchronized above.
+        return;
+    }
+
     size_t max_row_pixels =
         MIN(context_surfaces->source_pitch, context_surfaces->dest_pitch) /
         bytes_per_pixel;
@@ -154,7 +162,38 @@
         leftover_bytes = clipped_dest_size - consumed_bytes;
     }
 
+    bool gpu_copy = pg->surface_scale_factor > 1 && surf_src &&
+                    surf_src->color && !surf_src->swizzle &&
+                    surf_src->initialized && !surf_src->upload_pending &&
+                    surf_src->shape.color_format ==
+                        NV097_SET_SURFACE_FORMAT_COLOR_LE_A8R8G8B8 &&
+                    context_surfaces->color_format ==
+                        NV062_SET_COLOR_FORMAT_LE_A8R8G8B8 &&
+                    image_blit->operation == NV09F_SET_OPERATION_SRCCOPY &&
+                    !source_offset && !dest_offset &&
+                    image_blit->width == surf_src->width &&
+                    image_blit->height == surf_src->height &&
+                    context_surfaces->source_pitch == image_blit->width * 4 &&
+                    context_surfaces->dest_pitch == context_surfaces->source_pitch &&
+                    row_pixels == image_blit->width &&
+                    clipped_dest_size == dest_size &&
+                    (dest_addr + dest_size <= source_addr ||
+                     source_addr + surf_src->size <= dest_addr);
+
     SurfaceBinding *surf_dest = pgraph_vk_surface_get(d, dest_addr);
+    if (gpu_copy && !surf_dest) {
+        surf_dest = pgraph_vk_surface_create_blit_dest(d, surf_src,
+                                                       dest_addr);
+        gpu_copy = surf_dest != NULL;
+    }
+    if (gpu_copy && surf_dest) {
+        gpu_copy = surf_dest->color && !surf_dest->swizzle &&
+                   surf_dest->width == surf_src->width &&
+                   surf_dest->height == surf_src->height &&
+                   surf_dest->pitch == surf_src->pitch &&
+                   surf_dest->host_fmt.vk_format ==
+                       surf_src->host_fmt.vk_format;
+    }
     if (surf_dest) {
         if (adjusted_height < surf_dest->height ||
             row_pixels < surf_dest->width) {
@@ -224,4 +263,8 @@
                                    DIRTY_MEMORY_VGA);
     memory_region_set_client_dirty(d->vram, dest_addr, clipped_dest_size,
                                    DIRTY_MEMORY_NV2A_TEX);
+
+    if (gpu_copy) {
+        pgraph_vk_surface_copy_blit(d, surf_src, surf_dest);
+    }
 }
diff --git a/hw/xbox/nv2a/pgraph/vk/renderer.h b/hw/xbox/nv2a/pgraph/vk/renderer.h
--- a/hw/xbox/nv2a/pgraph/vk/renderer.h
+++ b/hw/xbox/nv2a/pgraph/vk/renderer.h
@@ -511,6 +511,11 @@
 void pgraph_vk_surface_flush(NV2AState *d);
 void pgraph_vk_process_pending_downloads(NV2AState *d);
 void pgraph_vk_surface_download_if_dirty(NV2AState *d, SurfaceBinding *surface);
+SurfaceBinding *pgraph_vk_surface_create_blit_dest(NV2AState *d,
+                                                   SurfaceBinding *source,
+                                                   hwaddr dest_addr);
+void pgraph_vk_surface_copy_blit(NV2AState *d, SurfaceBinding *source,
+                                 SurfaceBinding *dest);
 SurfaceBinding *pgraph_vk_surface_get_within(NV2AState *d, hwaddr addr);
 void pgraph_vk_wait_for_surface_download(SurfaceBinding *e);
 void pgraph_vk_download_dirty_surfaces(NV2AState *d);
diff --git a/hw/xbox/nv2a/pgraph/vk/surface.c b/hw/xbox/nv2a/pgraph/vk/surface.c
--- a/hw/xbox/nv2a/pgraph/vk/surface.c
+++ b/hw/xbox/nv2a/pgraph/vk/surface.c
@@ -682,6 +682,8 @@
     }
 }
 
+static void create_surface_image(PGRAPHState *pg, SurfaceBinding *surface);
+
 static void surface_put(NV2AState *d, SurfaceBinding *surface)
 {
     PGRAPHVkState *r = d->pgraph.vk_renderer_state;
@@ -692,6 +694,84 @@
     register_cpu_access_callback(d, surface);
 
     QTAILQ_INSERT_HEAD(&r->surfaces, surface, entry);
+}
+
+SurfaceBinding *pgraph_vk_surface_create_blit_dest(NV2AState *d,
+                                                   SurfaceBinding *source,
+                                                   hwaddr dest_addr)
+{
+    PGRAPHState *pg = &d->pgraph;
+    PGRAPHVkState *r = pg->vk_renderer_state;
+    SurfaceBinding *other;
+
+    QTAILQ_FOREACH(other, &r->surfaces, entry) {
+        if (dest_addr < other->vram_addr + other->size &&
+            other->vram_addr < dest_addr + source->size) {
+            return NULL;
+        }
+    }
+
+    SurfaceBinding *dest = g_new0(SurfaceBinding, 1);
+
+    dest->vram_addr = dest_addr;
+    dest->shape = source->shape;
+    dest->dma_addr = dest_addr;
+    dest->dma_len = source->size;
+    dest->color = true;
+    dest->width = source->width;
+    dest->height = source->height;
+    dest->pitch = source->pitch;
+    dest->size = source->size;
+    dest->fmt = source->fmt;
+    dest->host_fmt = source->host_fmt;
+    dest->frame_time = pg->frame_time;
+
+    create_surface_image(pg, dest);
+    surface_put(d, dest);
+    return dest;
+}
+
+void pgraph_vk_surface_copy_blit(NV2AState *d, SurfaceBinding *source,
+                                 SurfaceBinding *dest)
+{
+    PGRAPHState *pg = &d->pgraph;
+    VkCommandBuffer cmd = pgraph_vk_begin_nondraw_commands(pg);
+    VkFormat format = source->host_fmt.vk_format;
+
+    pgraph_vk_transition_image_layout(pg, cmd, source->image, format,
+                                      VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL,
+                                      VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL);
+    pgraph_vk_transition_image_layout(pg, cmd, dest->image, format,
+                                      VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL,
+                                      VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL);
+
+    VkImageCopy region = {
+        .srcSubresource.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT,
+        .srcSubresource.layerCount = 1,
+        .dstSubresource.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT,
+        .dstSubresource.layerCount = 1,
+        .extent.width = source->width * pg->surface_scale_factor,
+        .extent.height = source->height * pg->surface_scale_factor,
+        .extent.depth = 1,
+    };
+    vkCmdCopyImage(cmd, source->image, VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL,
+                   dest->image, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, 1,
+                   &region);
+
+    pgraph_vk_transition_image_layout(pg, cmd, source->image, format,
+                                      VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL,
+                                      VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL);
+    pgraph_vk_transition_image_layout(pg, cmd, dest->image, format,
+                                      VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL,
+                                      VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL);
+    pgraph_vk_end_nondraw_commands(pg, cmd);
+
+    dest->initialized = true;
+    dest->upload_pending = false;
+    dest->download_pending = false;
+    dest->draw_dirty = false;
+    dest->frame_time = pg->frame_time;
+    dest->draw_time = pg->draw_time;
 }
 
 SurfaceBinding *pgraph_vk_surface_get(NV2AState *d, hwaddr addr)
diff --git a/hw/xbox/nv2a/pgraph/vk/draw.c b/hw/xbox/nv2a/pgraph/vk/draw.c
--- a/hw/xbox/nv2a/pgraph/vk/draw.c
+++ b/hw/xbox/nv2a/pgraph/vk/draw.c
@@ -1445,9 +1445,14 @@
                      vp_height = pg->surface_binding_dim.height;
         pgraph_apply_scaling_factor(pg, &vp_width, &vp_height);
 
+        float vp_offset = pg->surface_scale_factor > 1
+                              ? 0.5f * (pg->surface_scale_factor - 1) + 0.25f
+                              : 0.0f;
         VkViewport viewport = {
-            .width = vp_width,
-            .height = vp_height,
+            .x = -vp_offset,
+            .y = -vp_offset,
+            .width = vp_width + vp_offset,
+            .height = vp_height + vp_offset,
             .minDepth = 0.0,
             .maxDepth = 1.0,
         };

**Comment by r3538987** [comment]
Update1. And first casualties - later in second mission/section, when you move camera towards region where are lasers or steam behind the wall. 

https://github.com/user-attachments/assets/ad45da85-006f-4a77-89ab-449cf1af419d

**Comment by r2bir2** [comment]
Either this @r3538987

Or my patch, works with almost every Version of xemu: https://www.reddit.com/r/Splintercell/comments/1wfe11s/double_agent_v2_green_border_antialiasing_fixed/
It has intact fog rendering from Update1

**Comment by r3538987** [comment]
> Either this [@r3538987](https://github.com/r3538987)
> 
> Or my patch, works with almost every Version of xemu: https://www.reddit.com/r/Splintercell/comments/1wfe11s/double_agent_v2_green_border_antialiasing_fixed/ It has intact fog rendering from Update1

Thanks for pointing this out, I'm likely reinventing the wheel.
I will try your patch.

**Comment by r3538987** [comment]
Updated patch. https://github.com/xemu-project/xemu/issues/949#issuecomment-5758885106
TLDR.
- Patch for Vulkan renderer only.
- Gets rid of glow issue on screen sides with higher scaling settings.
- Removes pixelization issue with 2x 3x... scaling values. (preserves hi-res image also in scenes with fog/steam/smoke).
- Keeps original bloom effect, despite it likely still broken with scaling cranked up. 

@r2bir2 People surely would appreciate various options with patches. Plus as mentioned on Reddit, night vision goggles still have screen edge glow. Probably fixable quite easily.

**Comment by Davometric** [comment]
> Either this [@r3538987](https://github.com/r3538987)
> 
> Or my patch, works with almost every Version of xemu: https://www.reddit.com/r/Splintercell/comments/1wfe11s/double_agent_v2_green_border_antialiasing_fixed/ It has intact fog rendering from Update1

This works great! I was honestly sick of using the ReShade toggle all the time! Do you have a patch for Chaos Theory too?
